torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <torch/nn/module.h>
+#include <torch/nn/modules/any.h>
+
+#include <ATen/Error.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace torch { namespace nn {
+
+/// A `Sequential` module is a container for any number of other modules. Its
+/// `forward()` method chains outputs to inputs and returns the final output.
+/// The `Sequential` class reference semantics.
+class Sequential : public CloneableModule<Sequential> {
+ public:
+  using Iterator = std::vector<std::shared_ptr<AnyModule>>::iterator;
+
+  /// Constructs the `Sequential` from a pack of modules. Each module can either
+  /// be a plain value (e.g. `Linear`) or a boxed value (e.g.
```

Comment:
No if you pass an lvalue (i.e. not a temporary) into `push_back`, it will just copy it into a newly created boxed instance internally. Since `Variable`/`Tensor` is itself refcounted, this copy would lead to shared ownership of variables. If you pass a temporary or `std::move` a module before passing it in, the `Sequential` will take ownership.

I need to think a bit harder about what ownership semantics modules should have in general: value, reference, copy etc.